File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <bits/stdc++.h>
using namespace std;

set<int> st;
string n;
bool vis[7];

bool chk(int num) {
	if (num <= 1) return false;
	for (int i = 2; i < num; i++)
		if (num % i == 0) return false;
	return true;
}

void dfs(string s) {
	if (s.size() > n.size()) return;

	int num = stoi(s);
	if (chk(num)) st.insert(num);

	for (int i = 0; i < n.size(); i++) {
		if (vis[i]) continue;

		vis[i] = true;
		dfs(s + n[i]);
		vis[i] = false;
	}
}

int solution(string numbers) {
	n = numbers;
	for (int i = 0; i < n.size(); i++) {
		string s = "";
		s += n[i];
		vis[i] = true;
		dfs(s);
		vis[i] = false;
	}
	return st.size();
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
vis = [False for _ in range(7)]
uniq = []

def check(s):
    conv = int(s)

    if conv <= 1:
        return

    for i in range(2, conv):
        if conv % i == 0:
            return

    uniq.append(conv)
    return

def DFS(s):
    if len(s) > len(num):
        return
    
    for i, n in enumerate(num):
        if vis[i] == False:
            vis[i] = True
            check(s + n)
            DFS(s + n)
            vis[i] = False


def solution(numbers : str):
    global num
    num = numbers
    answer = 0

    DFS("")

    return len(list(set(uniq)))